### DISCUSSION TITLE
JWT not working

### DISCUSSION THREAD

**Discussion by akashshyamdev** [nextauthjs/next-auth#2928] [Help] [answered]
Hey guys,

I'm using the following setup in `[...nextauth]`:

```js
import clientPromise from "@middleware/mongodb";
import { MongoDBAdapter } from "@next-auth/mongodb-adapter";
import NextAuth from "next-auth";
import FacebookProvider from "next-auth/providers/facebook";
import GoogleProvider from "next-auth/providers/google";
import TwitterProvider from "next-auth/providers/twitter";

export default NextAuth({
  adapter: MongoDBAdapter({
    db: (await clientPromise).db("dev"),
  }),
  providers: [
    FacebookProvider({
      clientId: process.env.FACEBOOK_CLIENT_ID,
      clientSecret: process.env.FACEBOOK_CLIENT_SECRET,
    }),
    GoogleProvider({
      clientId: process.env.GOOGLE_CLIENT_ID,
      clientSecret: process.env.GOOGLE_CLIENT_SECRET,
    }),
    TwitterProvider({
      clientId: process.env.TWITTER_CLIENT_ID,
      clientSecret: process.env.TWITTER_CLIENT_SECRET,
    }),
  ],
  callbacks: {
    async session({ session, token, user }) {
      console.log("session", session);
      console.log("user", user);
      console.log("token", token);

      session.id = user?.id;
      session.accessToken = token.accessToken;

      return session;
    },
  },
  secret: process.env.JWT_SECRET,
  session: {
    jwt: true,
  },
});
```

But for some reason, I don't get(in the frontend) a JWT token(a string) but instead an object containing `{user: {name, email, image}, expires: Date}`. All I want is the user's _ID(mongodb default unique ID) inside the session/JWT token. Am I missing something here?

**Comment by balazsorban44** [comment, marked as answer]
https://next-auth.js.org/configuration/callbacks#jwt-callback

https://next-auth.js.org/configuration/options#session

**Reply by akashshyamdev** [reply]
@balazsorban44 thanks for the links, but that doesn't really solve my problem of getting a mongodb/mongoose ID and the JWT. Could you provide a little more detail?

btw, I feel the docs can be improved with a quick start guide, and I'd love to do that later this week(don't really have time right now).

**Reply by balazsorban44** [reply]
When using a JWT session (which you do by setting `session.jwt` to `true` explicitly)

```js
jwt({ token, user }) {
  if(user) token.id = user.id // User is only available on sign-in
  return token
},
session({ session, token }) {
  session.user.id = token.id
  return session
}
```

Our callbacks documentation seems to be less visible/hard to understand(?) for some reason - even if I think they are the most powerful part of `next-auth` - so I am happy for any feedback on how to make it better.

**Reply by akashshyamdev** [reply]
I agree with you, callbacks seem very powerful in `next-auth`. Let's drop the JWT token thing, and just go with sessions. Is there a way to access the default `_id`(type - `ObjectId`) created by mongodb? This `_id` can be used in the future for references between two different collections.

**Reply by akashshyamdev** [reply]
Also, using a JWT with a session sort of defeats the point of a JWT, which is storing the user's auth state on the client instead of the server.

**Reply by balazsorban44** [reply]
yes, as shown above.

when using a jwt session (not a database persisted one), the token's content is determined by what you return in the jwt callback. the session received in the client is only a subset of that by default, for privacy/security reasons. if you absolutely need the ID client side, you can return that as well.

**Reply by balazsorban44** [reply]
you CAN use a database persisted session, just drop the session.jwt true config. When using an adapter, the default is to store the session in the database

**Reply by akashshyamdev** [reply]
Ok I've understood this, thanks a lot for your help! Are there any pending issues for me to help with? Also, I think there should be examples with typescript too. Secondly, how can I access the `user` in the `jwt` callback on signup?  Finally, how can I extend `session.user` to add a `id` with typescript?

**Reply by balazsorban44** [reply]
Feel free to suggest docs changes if you think you could improve it. You can open a PR to https://github.com/nextauthjs/docs 


We have TS docs at https://next-auth.js.org/getting-started/typescript#module-augmentation
